azure/create-spn.sh

```diff
@@ -135,24 +150,70 @@ show_account()
 
 create_ad_sp()
 {
-  echo "Creating Azure AD Service Principal and configuring its access to Azure resources in subscription ${subscription}..."
-  years="${years:-1}"
-  spn=$(az ad sp create-for-rbac -o json -n ${display_name} --role Owner --scopes /subscriptions/${subscription} --years ${years})
+  if [[ -z "${client_id}" ]]; then
+    echo "Creating Azure AD Service Principal and configuring its access to Azure resources in subscription ${subscription}..."
+    years="${years:-1}"
+    spn=$(az ad sp create-for-rbac -o json -n "${display_name}" --only-show-errors \
+      --role Owner --scopes /subscriptions/${subscription} --years ${years})
+  else
+    echo "Update application ${client_id} MS Graph API permissions"
+  fi
+  echo "Waiting for Azure AD Service Principal to propagate..."
+  sleep 15
+
+}
+
+grant_api_permissions()
+{
+  [[ -z "${client_id}" ]] && client_id=$(echo "${spn}" | jq -r .appId)
+  sp_object_id=$(az ad sp show --id "${client_id}" -o tsv --query objectId)
+  user_object_id=$(az ad signed-in-user show -o tsv --query objectId)
+
+  # To add owners to application
+  az ad app owner add --id "${client_id}" --owner-object-id "${sp_object_id}"
+  az ad app owner add --id "${client_id}" --owner-object-id "${user_object_id}"
+
+  # To add owners to service principal
+  az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/${sp_object_id}/owners/\$ref \
+    --headers Content-Type=application/json \
+    -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/servicePrincipals/${sp_object_id}\"}"
+  az rest -m POST -u https://graph.microsoft.com/beta/servicePrincipals/${sp_object_id}/owners/\$ref \
+    --headers Content-Type=application/json \
+    -b "{\"@odata.id\": \"https://graph.microsoft.com/beta/users/${user_object_id}\"}"
+
+  # To add API permissions
+  # Application.ReadWrite.OwnedBy
+  az ad app permission add --id "${client_id}" --only-show-errors \
+    --api 00000003-0000-0000-c000-000000000000 \
```

Comment:
these ids are magical, 
can we add some comments/information about what these magic ids mean

Reply:
add comments for API permissions.